### ISSUE TITLE
TypeScript error when using two different response types (error or success) using c.jsonT

### ISSUE THREAD

**Issue by banjo** [honojs/hono#1377]
### What version of Hono are you using?

3.5.4

### What runtime/platform is your app running on?

Node

### What steps can reproduce the bug?

Create a simple route, in one case it should return one type (in my case, an error), and in another case, some other data (which is the actual data in my case). 

These are the types:

```
type Success<T> = {
    success: true;
    data: T;
};

type Error = {
    success: false;
    message: string;
};

type ResultType<T> = Success<T> | Error;
```

This is the simplest possible example:

```
endpoint.get("/all", c => {
    const data = getData();

    if (!data.success) {
        return c.jsonT({
            success: false,
            message: "Could not fetch data"
        });
    }

    return c.jsonT({
        success: true,
        data: data.data,
    });
});
```

In this example, it renders this very long error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '(c: Context<{ Variables: AuthVariables; }, "/all", {}>) => TypedResponse<{ success: boolean; message: string; status: number; }> | TypedResponse<{ success: boolean; data: {}; }>' is not assignable to parameter of type 'H<{ Variables: AuthVariables; }, "/all", {}, { success: boolean; message: string; status: number; }>'.
      Type '(c: Context<{ Variables: AuthVariables; }, "/all", {}>) => TypedResponse<{ success: boolean; message: string; status: number; }> | TypedResponse<{ success: boolean; data: {}; }>' is not assignable to type 'MiddlewareHandler<{ Variables: AuthVariables; }, "/all", {}>'.
        Type 'TypedResponse<{ success: boolean; message: string; status: number; }> | TypedResponse<{ success: boolean; data: {}; }>' is not assignable to type 'Promise<void | Response>'.
          Type 'TypedResponse<{ success: boolean; message: string; status: number; }>' is missing the following properties from type 'Promise<void | Response>': then, catch, finally, [Symbol.toStringTag]ts(2769)
```

### What is the expected behavior?

I would like to be able to return two different json bodies depending on the circumstances, and still be able to get type safety.

### What do you see instead?

I'm not sure for now, thinking about using normal c.json for the error response, but that would probably mean I would lose some type safety I presume.

**Comment by yusukebe** [comment]
Hi @banjo !

This is a known issue. I believe it's quite challenging to address, but I will investigate how we can support a solution. Thanks!

**Comment by yusukebe** [comment]
@banjo

I've just released version v3.5.5! This should resolve the issue you mentioned. Please give it a try, and if you encounter any problems, feel free to reopen the issue.
